db/migrate/20221216171738_value_overrides_to_non_canonical_profiles.rb

```diff
@@ -0,0 +1,13 @@
+class ValueOverridesToNonCanonicalProfiles < ActiveRecord::Migration[7.0]
+  class Profile < ApplicationRecord; end
+
+  def up
+    sub_query = Profile.where('p.parent_profile_id = id').select(:value_overrides).to_sql
+    query = "UPDATE profiles AS p SET value_overrides = (#{sub_query}) WHERE parent_profile_id IS NOT NULL"
+    ActiveRecord::Base.connection.execute(query)
```

Comment:
This looks a little weird but I had to do it like this because `joins` and `update_all` don't work correctly together: https://github.com/rails/rails/issues/522

[This comment](https://github.com/rails/rails/issues/522#issuecomment-426037101) explains the issue and they came up with this workaround that I used

Reply:
You should not write the whole query as a string, only put the string into `update_all` part of the original one.

Reply:
I've tried this: 
`Profile.canonical(false).joins(:parent_profile).update_all('value_overrides = parent_profiles_profiles.value_overrides')`

But it returns this error:
```
ERROR:  missing FROM-clause entry for table "parent_profiles_profiles"
LINE 1: UPDATE "profiles" SET value_overrides = parent_profiles_prof...
```

The error made sense after looking at the SQL query that's generated from this:
`UPDATE "profiles" 
SET value_overrides = parent_profiles_profiles.value_overrides
WHERE "profiles"."id" IN (SELECT "profiles"."id" FROM "profiles" INNER JOIN "profiles" "parent_profiles_profiles" ON "parent_profiles_profiles"."id" = "profiles"."parent_profile_id" WHERE "profiles"."parent_profile_id" IS NOT NULL)`

`parent_profiles_profiles` is in a nested query and can't be accessed from where the query is trying to use it. This discussion talks about the issue: https://github.com/rails/rails/issues/522

Reply:
Try using an explicit join: `joins("parent_profile pp ON pp.id = profiles.parent_profile_id")` and then refer to it as `pp` in the update.

Reply:
I tried using an explicit join but had to update it slightly because `parent_profile` table doesn't exist so I used `profiles`: 
`Profile.canonical(false).joins("join profiles pp ON pp.id = profiles.parent_profile_id").update_all('value_overrides = pp.value_overrides')`

But got the same error 🙁 
```
Caused by PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "pp"
LINE 1: UPDATE "profiles" SET value_overrides = pp.value_overrides W...
```

The join is inside of a nested query again so that's why it's not working:
```
UPDATE "profiles" 
SET value_overrides = pp.value_overrides 
WHERE "profiles"."id" IN (SELECT "profiles"."id" FROM "profiles" join 
parent_profiles pp ON pp.id = profiles.parent_profile_id WHERE
 "profiles"."parent_profile_id" IS NOT NULL)
```

Reply:
You'd need more explicit:
```suggestion
    update_all('profiles.value_overrides = pp.value_overrides')
```
But it might not work still...

Reply:
Anyway, there's always arel ...

Reply:
`Profile.canonical(false).joins("join parent_profiles pp ON pp.id = profiles.parent_profile_id").update_all('profiles.value_overrides = pp.value_overrides')` still doesn't work 🙁 

I'll try to use arel instead!